### ISSUE TITLE
IEnumerable<FileEntry> returned from Extractor.Extract is actually a List<FileEntry>

### ISSUE THREAD

**Issue by andreas-eriksson** [microsoft/RecursiveExtractor#97]
I noticed that the Extract method stores all entries in a List before returning the result instead of yielding them one by one, as is the case in ExtractAsync.

I think this could lead to problems if the archive contains lots of entries as it will keep them all in memory.
Is there a reason to not using the yield pattern?

**Comment by gfs** [comment]
@andreas-eriksson This is because of the try block on like 593 of Extractor.cs. You can't yield out of a try.

https://github.com/microsoft/RecursiveExtractor/blob/0208151ece53b7ecfc4ce1bb20e82ee5dc7c8e86/RecursiveExtractor/Extractor.cs#L593-L610

It may be feasible to refactor to remove the try block however.

**Comment by andreas-eriksson** [comment]
Thanks, I noticed that as well. I may have a go at refactoring it a bit.

My real problem is that I am dealing with archives that contain millions of files and that are several GB in size and with lots of recursion. Would it be of interest if I looked at a solution that doesn't use temp files or copies files into memory streams?

I am thinking that it may be possible to let the consumer just get the stream directly in each FileEntry, by calling for instance zipFile.GetInputStream(zipEntry) on demand instead of directly. That could also be beneficiary in scenarios where the consumer is only interested in counting files or looking at file info.

**Comment by gfs** [comment]
I do plan to merge that suggestion for changing the temp files names either this weekend or early next week if that will help resolve.

From looking at the current code, the async code path is currently a truer enumeration, so that may be something you could try first, and if that helps it’s an indication we should switch to true enumeration on the sync api.

The reason we have to populate every stream ahead of time is that we need to check if each file itself is another archive, which would then need to be extracted as well (the recursive part of the extractor). I’ve also experienced issues with using the streams from the dependent libs directly because they will get passively disposed as you continue enumeration. thus the design with the full content stream.

if you can think of a solution that doesn’t require memory stream or delete on close file streams im willing to consider it for sure.

**Comment by andreas-eriksson** [comment]
Thanks, I'll see if I can make something work.

The current implementation of holding bytes in FileEntry.Content is a bit risky as it is to not notice that streams are not disposed until the enumerator is disposed. This is only an issue with archives with very many files and can be mitigated by manually disposing each FileEntry after processing it. It might be worth updating the examples on the main page to dispose it in the loop.

**Comment by gfs** [comment]
The FileEntty deconstructor will dispose the content stream automatically unless you have manually changed `DisposeOnFinalize` (for scenarios where you would want a file entry representation you created from a stream you may want to reuse).

I believe letting the reference leave context should be sufficient and the garbage collector should take care of it.

Are if you are finding the deconstructor isn’t firing properly? If so, a solution may be to have FileEntry implement IDisposable and have its dispose call the dispose of the stream.

https://github.com/microsoft/RecursiveExtractor/blob/8bfee99ca3edb2d3fb40f849f302c14a0d7644d1/RecursiveExtractor/FileEntry.cs#L204

**Comment by gfs** [comment]
I tried a quick implementation of this and it does seem to have changed some tested behaviors, so it seems not quite as trivial as I'd hoped, but I'll continue to investigate.

**Comment by andreas-eriksson** [comment]
FileEntries/Content is not out of scope until leaving the Extract method unless being explicitly disposed inside the foreach.
This will only be a problem for archives with a lot of files below the MemoryStreamCutoff.

```c#
using Microsoft.CST.RecursiveExtractor;

private void Extract()
{
    var path = "path/to/file";
    var extractor = new Extractor();
    foreach(var file in extractor.Extract(path))
    {
        doSomething(file.Content); 
        file.Content.Dispose();
    }
}
```
It would probably be harder to make mistakes if the code could be used as below, but I doi uncerstand that this is much harder to implement.
```c#
using Microsoft.CST.RecursiveExtractor;

private void Extract()
{
    var path = "path/to/file";
    var extractor = new Extractor();
    foreach(var file in extractor.Extract(path))
    {
        using(var ms = file.GetContentStream())
        {
            doSomething(ms); 
        }
    }
}
```

**Comment by gfs** [comment]
@andreas-eriksson I think the latter behavior you want works now with the changes I made in #101.

I added this test case for this scenario, which I believe confirms that this using + automatic disposal mechanism works fine while enumerating.

```csharp
[DataTestMethod]
[DataRow("TestDataArchivesNested.Zip", 54, true)]
[DataRow("TestDataArchivesNested.Zip", 54, false)]
public void ExtractArchiveAndDisposeTest(string fileName, int expectedNumFiles, bool parallel)
{
    var extractor = new Extractor();
    var path = Path.Combine(Directory.GetCurrentDirectory(), "TestData", "TestDataArchives", fileName);
    var results = extractor.Extract(path, new ExtractorOptions(){Parallel = parallel});
    var disposedResults = new List<FileEntry>();
    foreach(var file in extractor.Extract(path))
    {
        disposedResults.Add(file);
        using var theStream = file.Content;
        // Do something with the stream.
        _ = theStream.ReadByte();
    }
    Assert.AreEqual(expectedNumFiles, disposedResults.Count);
    foreach (var disposedResult in disposedResults)
    {
        Assert.ThrowsException<ObjectDisposedException>(() => disposedResult.Content.Position);
    }
}
```

**Comment by gfs** [comment]
Thanks for the suggestion here. I have also modified the `ExtractToDirectory` methods using this method to dispose the already extracted streams after extraction.

**Comment by gfs** [comment]
This behavior now be available in the just released 1.2 packages. Thanks for the feedback, and please let me know if you encounter any other issues.
